[[100 ошибок в Go и как их избежать]]

Рассмотрим пример неправильной передачи контекста:
```go
func handler(w http.ResponseWriter, r *http.Request) {
	response, err := doSomeWork(r.Context(), r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	go func() {
		err := publish(r.Context(), response)
		// work
	}()
	writeResponse(response)
}
```
Есть хендлер, который выполняет некоторую задачу и возвращает ответ. Но прежде чем вернуть ответ, его нужно отправить в kafka. 

Мы должны знать, что http контекст отменяется при следующих случаях:
- Когда соединение клиента закрывается
- В случае HTTP/2, когда он оказывается отмененным
- Когда отбыл был направлен клиенту

Первый два пункта ясны. Но при втором возникает состояние гонки:
- Если ответ был отправлен после публикации в Kafka, то мы и возвращаем ответ и успешно публикуем сообщение
- Если ответ клиенту отправлен до или во время публикации Kafka, то сообщение не должно быть опубликовано

Проблема решается тем, что бы не передавть родительский контекст, а вызвать пустой контекст Background. Но что если контекст http содержит какие-то данные? Тогда нам нужно реализовать собственный контекст
```go
type Context interface {
	Deadline() (deadline time.Time, ok bool)
	Done() <- chan struct{}
	Err() error
	Value(key any) any
}
```
```go
// Пользовательская обертка как оберетка над исходным контекстом
type detach struct {
	ctx context.Context
}
func (d detach) Deadline (time.Time, bool) {return time.Time{}, false}
func (d detach) Done() <- chan struct{} {return nil}
func (d detach) Err() error {return nil}
// Делегируем вызов get value родительскому контексту
func (d detach) Value(key any) any {return d.ctx.Value(key)}
```
Кроме последнего метода все остальные возвращают значения по умолчанию, поэтому контекст никогда не считается просроченным или отмененным 